### PR TITLE
DAOS-10348 object: clone sub requests for CPD RPC

### PR DESCRIPTION
**Pull request by Nasf-Fan**
For OSA case, more than one object shards may reside on the same
DAOS target. So daos_cpd_disp_ent::dcde_write_cnt may be not the
real count of update sub requests. When dispatch CPD RPC, we need
to clone related sub requests to avoid sending some irrelevant sub
request to the remote DAOS target. During the clone operation, we
need to avoid copying the same sub request that are for the object
different but co-located shards repeatedly.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

**Comment by Nasf-Fan** [comment]
test_daos_degraded_mode failed for DAOS-10125, will be retested.

**Review by jolivier23 — DISMISSED** [review]

**Comment by Nasf-Fan** [comment]
@jolivier23 @wangdi1 @gnailzenh , please help to review the patch. Thanks!

**Review by jolivier23 — APPROVED** [review]

File: src/object/obj_rpc.h [hunk 1]
```diff
@@ -158,7 +158,7 @@ enum obj_rpc_flags {
 	ORF_CPD_LEADER		= (1 << 9),
 	/* Bulk data transfer for CPD RPC. */
 	ORF_CPD_BULK		= (1 << 10),
-	/* Contain EC split req, only used on CPD leader locally. */
+	/* Contain EC split req, only used on CPD leader locally. Obsolete - DAOS-10348. */
```
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/object/obj_rpc.h:161]: I will update the comment when refresh the patch next time.

File: src/object/obj_tx.c [hunk 2]
```diff
@@ -1218,6 +1222,11 @@ dc_tx_classify_common(struct dc_tx *tx, struct daos_cpd_sub_req *dcsr,
 			  shard->do_target_id, dtrg_nr);
 
 		dtrg = &dtrgs[shard->do_target_id];
+
+		dtrg->dtrg_flags |= tgt_flags;
+		if (unlikely(shard->do_shard != idx))
```
Comment (liuxuezhao) [inline src/object/obj_tx.c:1227]: I don't understand the related code, just confirm here the "do_shard != idx" is which case? OSA added temporary shard? thx

File: src/object/srv_obj_remote.c [hunk 3]
```diff
@@ -483,22 +505,22 @@ ds_obj_cpd_dispatch(struct dtx_leader_handle *dlh, void *arg, int idx,
 	if (dcde_parent == NULL)
 		D_GOTO(out, rc = -DER_INVAL);
 
-	count = dcde_parent->dcde_read_cnt + dcde_parent->dcde_write_cnt;
-	if (count < total || (exec_arg->flags & ORF_HAS_EC_SPLIT)) {
-		rc = ds_obj_cpd_clone_reqs(shard_tgt, dcde_parent,
-					   dcsr_parent, total, &dcde, &dcsr);
-		if (rc != 0)
+	if (shard_tgt->st_flags & DTF_REASSEMBLE_REQ) {
```
Comment (liuxuezhao) [inline src/object/srv_obj_remote.c:508]: the DTF_REASSEMBLE_REQ seems not really related EC IOD/sgl reassemble?

File: src/object/obj_tx.c (hunk 2 above)
Comment (Nasf-Fan) [reply src/object/obj_tx.c:1227]: Yes, it is for OSA case.

**Comment by Nasf-Fan** [comment]
Ping @daos-stack/daos-gatekeeper , thanks!

**Comment by Nasf-Fan** [comment]
Ping @daos-stack/daos-gatekeeper , thanks!

**Comment by Nasf-Fan** [comment]
Ping @daos-stack/daos-gatekeeper , thanks!

**Comment by jolivier23** [comment]
@Nasf-Fan apologies as master branch has been mostly closed for the last month.   The last refresh of this patch was 18 days ago and we are prioritizing 2.2 and 2.0 issues.  Can you please merge latest and refresh?

**Comment by Nasf-Fan** [comment]
> @Nasf-Fan apologies as master branch has been mostly closed for the last month. The last refresh of this patch was 18 days ago and we are prioritizing 2.2 and 2.0 issues. Can you please merge latest and refresh?

Yes, merged with the latest master.

**Pull request merged**